### DISCUSSION TITLE
How does a local build of this repo get picked and used over the default RTM build?

### DISCUSSION THREAD

**Discussion by IanKemp** [dotnet/aspnetcore#27207] [Q&A] [answered]
I have an ASP.NET Core project that is built against the RTM bits of ASP.NET Core 3.1.9 that were downloaded as part of the Visual Studio 16.7.6 update.

Now I want to make my own local build of ASP.NET Core 3.1.9, with changes to its source code, and have my project use that instead. Assume that I am able to do this and install the resultant artifacts successfully.

**How do I tell my project to use this local build instead of the default specified by `Microsoft.NET.Sdk.Web`?**

**Comment by Pilchie** [comment]
@dotnet/aspnet-build ?

**Comment by dougbu** [comment, marked as answer]
@IanKemp the following should work
1. If necessary, increase [`$(AspNetCorePatchVersion)` in eng/Versions.props](https://github.com/dotnet/aspnetcore/blob/24d335f5caff84eff87b58f55d1f5c11a0efb721/eng/Versions.props#L11) to something higher than what's available on NuGet.org e.g. `10` for `3.1.x` at the moment
2. `.\build.cmd -pack` or `./build.sh -pack` in the dotnet/aspnetcore repo
    - May need `-all` if C++ (ANCM) or Java packages are important in your scenario
3. `. .\activate.ps1` or `. ./activate.sh` in the dotnet/aspnetcore repo
    - This does interesting things such as pick up the targeting pack and shared framework layouts the build creates under the .dotnet/ folder
4. Add the `$(RepoRoot)artifacts/packages/$(Configuration)/Shipping` folder from the dotnet/aspnetcore as a feed in the NuGet.config file for your project e.g.
    ``` xml
    <add key="locals" value="c:\repos\aspnetcore\artifacts\packages\Debug\Shipping" />
    ```
5. Build your project

**Reply by IanKemp** [reply]
@dougbu I did all that before as per `BuildFromSource.md` - what I didn't realise was that my project **is** building using the 3.1.109 SDK I have under `C:\aspnetcore`, but for some reason it seems determined to map `Microsoft.NET.Sdk.Web` to version 3.1.8, not version 3.1.10 that I'm building.

I was able to figure this out by deleting the `%userprofile%\.nuget\packages\microsoft.aspnetcore.app.ref\3.1.8` directory, then changing my solution's `NuGet.config` to have my locally-built ASP.NET Core packages as the **only** source. The result is that I then got the following:

```
1>C:\not\important\my-asp-net-core-web-project.csproj : error NU1102: Unable to find package Microsoft.AspNetCore.App.Ref with version (= 3.1.8)
1>C:\not\important\my-asp-net-core-web-project.csproj : error NU1102:   - Found 1 version(s) in MyBuildOfAspNetCore [ Nearest version: 3.1.10 ]
```

(`MyBuildOfAspNetCore` is the name of the NuGet package source for my local build of ASP.NET Core - copy-pasted from `BuildFromSource.md`.)

Then I got a little smarter and turned the MSBuild verbosity up to the max, and found this:

```
1>  Using "ProcessFrameworkReferences" task from assembly "C:\aspnetcore\.dotnet\sdk\3.1.109\Sdks\Microsoft.NET.Sdk\targets\..\tools\net472/Microsoft.NET.Build.Tasks.dll".
1>  Task "ProcessFrameworkReferences"
...
1>    Task Parameter:
1>        KnownFrameworkReferences=
1>            Microsoft.NETCore.App
1>                    DefaultRuntimeFrameworkVersion=3.1.0
1>                    IsTrimmable=true
1>                    LatestRuntimeFrameworkVersion=3.1.9
1>                    RuntimeFrameworkName=Microsoft.NETCore.App
1>                    RuntimePackNamePatterns=Microsoft.NETCore.App.Runtime.**RID**
1>                    RuntimePackRuntimeIdentifiers=linux-arm;linux-arm64;linux-musl-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;tizen.4.0.0-armel;tizen.5.0.0-armel;win-arm;win-arm64;win-x64;win-x86
1>                    TargetFramework=netcoreapp3.1
1>                    TargetingPackName=Microsoft.NETCore.App.Ref
1>                    TargetingPackVersion=3.1.0
...
1>            Microsoft.AspNetCore.App
1>                    DefaultRuntimeFrameworkVersion=3.1.0
1>                    LatestRuntimeFrameworkVersion=3.1.9
1>                    RuntimeFrameworkName=Microsoft.AspNetCore.App
1>                    RuntimePackNamePatterns=Microsoft.AspNetCore.App.Runtime.**RID**
1>                    RuntimePackRuntimeIdentifiers=win-x64;win-x86;win-arm;osx-x64;linux-musl-x64;linux-musl-arm64;linux-x64;linux-arm;linux-arm64
1>                    TargetFramework=netcoreapp3.1
1>                    TargetingPackName=Microsoft.AspNetCore.App.Ref
1>                    TargetingPackVersion=3.1.8
...
1>    Output Item(s): 
1>        _PackageToDownload=
1>            Microsoft.AspNetCore.App.Ref
1>                    Version=3.1.8
...
1>    Output Item(s): 
1>        TargetingPack=
1>            Microsoft.NETCore.App
1>                    PackageConflictPreferredPackages=Microsoft.NETCore.App.Ref;Microsoft.NETCore.App.Runtime.linux-arm;Microsoft.NETCore.App.Runtime.linux-arm64;Microsoft.NETCore.App.Runtime.linux-musl-arm64;Microsoft.NETCore.App.Runtime.linux-musl-x64;Microsoft.NETCore.App.Runtime.linux-x64;Microsoft.NETCore.App.Runtime.osx-x64;Microsoft.NETCore.App.Runtime.rhel.6-x64;Microsoft.NETCore.App.Runtime.tizen.4.0.0-armel;Microsoft.NETCore.App.Runtime.tizen.5.0.0-armel;Microsoft.NETCore.App.Runtime.win-arm;Microsoft.NETCore.App.Runtime.win-arm64;Microsoft.NETCore.App.Runtime.win-x64;Microsoft.NETCore.App.Runtime.win-x86
1>                    PackageDirectory=C:\aspnetcore\.dotnet\packs\Microsoft.NETCore.App.Ref\3.1.0
1>                    PackageName=Microsoft.NETCore.App.Ref
1>                    PackageVersion=3.1.0
1>                    Path=C:\aspnetcore\.dotnet\packs\Microsoft.NETCore.App.Ref\3.1.0
1>                    RuntimeFrameworkName=Microsoft.NETCore.App
1>                    RuntimePackRuntimeIdentifiers=linux-arm;linux-arm64;linux-musl-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;tizen.4.0.0-armel;tizen.5.0.0-armel;win-arm;win-arm64;win-x64;win-x86
1>                    TargetFramework=netcoreapp3.1
1>                    TargetingPackFormat=
...
1>            Microsoft.AspNetCore.App
1>                    PackageConflictPreferredPackages=Microsoft.AspNetCore.App.Ref;Microsoft.AspNetCore.App.Runtime.win-x64;Microsoft.AspNetCore.App.Runtime.win-x86;Microsoft.AspNetCore.App.Runtime.win-arm;Microsoft.AspNetCore.App.Runtime.osx-x64;Microsoft.AspNetCore.App.Runtime.linux-musl-x64;Microsoft.AspNetCore.App.Runtime.linux-musl-arm64;Microsoft.AspNetCore.App.Runtime.linux-x64;Microsoft.AspNetCore.App.Runtime.linux-arm;Microsoft.AspNetCore.App.Runtime.linux-arm64
1>                    PackageName=Microsoft.AspNetCore.App.Ref
1>                    PackageVersion=3.1.8
1>                    RuntimeFrameworkName=Microsoft.AspNetCore.App
1>                    RuntimePackRuntimeIdentifiers=win-x64;win-x86;win-arm;osx-x64;linux-musl-x64;linux-musl-arm64;linux-x64;linux-arm;linux-arm64
1>                    TargetFramework=netcoreapp3.1
1>                    TargetingPackFormat=
```

While the `ProcessFrameworkReferences` task finds my build of `Microsoft.NETCore.App` at `C:\aspnetcore\.dotnet\packs\Microsoft.NETCore.App.Ref\3.1.0`, it **does not** find my build of `Microsoft.AspNetCore.App` at `C:\aspnetcore\.dotnet\packs\Microsoft.AspNetCore.App.Ref\3.1.10` - it wants to download the 3.1.8 refs from NuGet (into `%userprofile%\.nuget\packages\microsoft.aspnetcore.app.ref\3.1.8`).

After spending a couple of hours trying to figure out the "why" behind the above in order to fix it, I gave up and just renamed `C:\aspnetcore\.dotnet\packs\Microsoft.AspNetCore.App.Ref\3.1.10` to `C:\aspnetcore\.dotnet\packs\Microsoft.AspNetCore.App.Ref\3.1.8`.

And it works!

I'm not sure if my environment is broken somewhere somehow, or if there's a bug in `ProcessFrameworkReferences`, or what, but for now I can actually continue investigating what I wanted to at the start of all this, so thanks again.

**Comment by dougbu** [comment]
FYI setting `<RuntimeFrameworkVersion>3.1.10</RuntimeFrameworkVersion>` might have saved you the rename.

**Reply by IanKemp** [reply]
> FYI setting `<RuntimeFrameworkVersion>3.1.10</RuntimeFrameworkVersion>` might have saved you the rename.

Setting that gives an MSBuild error:

```
Unable to find package Microsoft.NETCore.App.Host.win-x64. No packages exist with this id in source(s): MyBuildOfAspNetCore
```

And indeed, `C:\aspnetcore\.dotnet\packs\Microsoft.NETCore.App.Host.win-x64` does not contain a `3.1.10` directory (but does contain a `3.1.9` directory). `.\build.cmd -all` doesn't generate that directory either, so not sure what else I'm missing.

**Comment by dougbu** [comment]
Oops, I got confused between NETCore.App and AspNetCore.App. Sorry for wasting your time.

**Reply by IanKemp** [reply]
Not a problem, I appreciate the help so far. I'm aware that my timing is not the best, as you're probably rather busy with finishing up work for the very imminent launch of .NET 5.

If you do have any more time assisting me, and are willing to spend that time, and are aware of any reason off the top of your head why the build process would be choosing 3.1.8 instead of 3.1.10, please can you let me know - I'd really like to **understand** what's going on, rather than just cargo-cult do the folder rename to make everything work.

Otherwise I will play with the relevant source from `dotnet/sdk` and see if I can figure it out.
